cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -0,0 +1,215 @@
+        lastMessageId = message.getId();
+        controlResponses.add(new TetheringControlResponse(lastMessageId, controlMessage));
+      }
+    } catch (TopicNotFoundException e) {
+      LOG.warn("Received control connection from peer {} that's not tethered", peer);
+    } catch (IllegalArgumentException e) {
+      throw new BadRequestException(String.format("Invalid message id %s", messageId));
+    }
+
+    if (controlResponses.isEmpty()) {
+      controlResponses.add(new TetheringControlResponse(
+        lastMessageId, new TetheringControlMessage(TetheringControlMessage.Type.KEEPALIVE)));
+    }
+    responder.sendJson(HttpResponseStatus.OK,
+                       GSON.toJson(controlResponses.toArray(new TetheringControlResponse[0]),
+                                   TetheringControlResponse[].class));
+  }
+
+  /**
+   * Creates a tethering with a client.
+   */
+  @PUT
+  @Path("/tethering/connect")
+  public void createTethering(FullHttpRequest request, HttpResponder responder)
+    throws NotImplementedException, IOException {
+    checkTetheringServerEnabled();
+
+    String content = request.content().toString(StandardCharsets.UTF_8);
+    TetheringConnectionRequest tetherRequest = GSON.fromJson(content, TetheringConnectionRequest.class);
+    TopicId topicId = new TopicId(NamespaceId.SYSTEM.getNamespace(),
+                                  topicPrefix + tetherRequest.getPeer());
+    try {
+      messagingService.createTopic(new TopicMetadata(topicId, Collections.emptyMap()));
+    } catch (TopicAlreadyExistsException e) {
+      LOG.warn("Topic {} already exists", topicId);
+    } catch (IOException e) {
+      LOG.error("Failed to create topic {}", topicId, e);
+      throw e;
+    }
+
+    // We don't need to keep track of the client metadata on the server side.
+    PeerMetadata peerMetadata = new PeerMetadata(tetherRequest.getNamespaceAllocations(), Collections.emptyMap());
+    // We don't store the peer endpoint on the server side because the connection is initiated by the client.
+    PeerInfo peerInfo = new PeerInfo(tetherRequest.getPeer(), null, TetheringStatus.PENDING, peerMetadata);
+    try {
+      store.addPeer(peerInfo);
+    } catch (PeerAlreadyExistsException pae) {
+      // Peer is already configured, treat this as a no-op.
+      responder.sendStatus(HttpResponseStatus.OK);
+      return;
+    } catch (Exception e) {
+      try {
+        messagingService.deleteTopic(topicId);
+      } catch (Exception ex) {
+        e.addSuppressed(ex);
+      }
+      throw new IOException("Failed to create tethering with peer " + tetherRequest.getPeer(), e);
+    }
+    responder.sendStatus(HttpResponseStatus.OK);
+  }
+
+  /**
+   * Accepts/rejects the tethering request.
+   */
+  @POST
+  @Path("/tethering/connections/{peer}")
+  public void tetherAction(HttpRequest request, HttpResponder responder, @PathParam("peer") String peer,
+                           @QueryParam("action") String action)
```

Comment:
Typically for POST call, the request parameters should be in the body, not as query parameter. Better to define a request object and have the action encapsulated inside.

Reply:
done